src/extension-runners/firefox-desktop.js

```diff
@@ -0,0 +1,229 @@
+    }
+
+    this.runningInfo.firefox.kill();
+  }
+
+  // Private methods.
+
+  async setupProfileDir() {
+    const {
+      customPrefs,
+      extensions,
+      keepProfileChanges,
+      preInstall,
+      profilePath,
+      firefoxApp,
+    } = this.params;
+
+    if (profilePath) {
+      if (keepProfileChanges) {
+        log.debug(`Using Firefox profile from ${profilePath}`);
+        this.profile = await firefoxApp.useProfile(profilePath, {customPrefs});
+      } else {
+        log.debug(`Copying Firefox profile from ${profilePath}`);
+        this.profile = await firefoxApp.copyProfile(profilePath, {customPrefs});
+      }
+    } else {
+      log.debug('Creating new Firefox profile');
+      this.profile = await firefoxApp.createProfile({customPrefs});
+    }
+
+    // preInstall the extensions if needed.
+    if (preInstall) {
+      for (const extension of extensions) {
+        await firefoxApp.installExtension({
+          asProxy: true,
+          extensionPath: extension.sourceDir,
+          manifestData: extension.manifestData,
+          profile: this.profile,
+        });
+      }
+    }
+  }
+
+  async startFirefoxInstance() {
+    const {
+      browserConsole,
+      extensions,
+      firefoxBinary,
+      preInstall,
+      startUrl,
+      firefoxApp,
+      firefoxClient,
+    } = this.params;
+
+    const binaryArgs = [];
+
+    if (browserConsole) {
+      binaryArgs.push('-jsconsole');
+    }
+    if (startUrl) {
+      const urls = Array.isArray(startUrl) ? startUrl : [startUrl];
+      for (const url of urls) {
+        binaryArgs.push('--url', url);
+      }
+    }
+
+    this.runningInfo = await firefoxApp.run(this.profile, {
+      firefoxBinary, binaryArgs,
+    });
+
+    this.runningInfo.firefox.on('close', () => {
+      for (const fn of this.cleanupCallbacks) {
+        try {
+          fn();
+        } catch (error) {
+          log.debug(`Exception on executing cleanup callback: ${error}`);
+        }
+      }
+    });
+
+    if (!preInstall) {
+      const remoteFirefox = this.remoteFirefox = await firefoxClient({
+        port: this.runningInfo.debuggerPort,
+      });
+
+      // Install all the temporary addons.
+      for (const extension of extensions) {
+        try {
+          const addonId = await (
+            remoteFirefox.installTemporaryAddon(extension.sourceDir)
+              .then((installResult: FirefoxRDPResponseAddon) => {
+                return installResult.addon.id;
+              })
+          );
+
+          if (!addonId) {
+            throw new WebExtError(
+              'Unexpected missing addonId in the installAsTemporaryAddon result'
+            );
+          }
+
+          this.reloadableExtensions.set(extension.sourceDir, addonId);
+        } catch (error) {
+          if (error instanceof RemoteTempInstallNotSupported) {
+            log.debug(`Caught: ${error}`);
+            throw new WebExtError(
```

Comment:
Hmm, I wonder why this wasn't a `UsageError` before. Probably nothing to worry about in this patch.